diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7b26c03..f7b67d2 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -322,6 +322,7 @@ SET(CUDA_SOURCES
         src/hardware/gpu/nn/gpu_pool.cu
         src/hardware/gpu/nn/gpu_pool_kernels.cu
         src/hardware/gpu/nn/gpu_tensor_nn.cu
+        src/hardware/gpu/nn/gpu_tensor_nn_kernels.cu
         )
 SET(SOURCES ${CPP_SOURCES} ${CUDA_SOURCES})
 
